compliance/model/src/test/java/org/eclipse/rdf4j/model/util/IsomorphicBenchmark.java

```diff
@@ -0,0 +1,123 @@
+package org.eclipse.rdf4j.model.util;
+
+import org.eclipse.rdf4j.model.Model;
+import org.eclipse.rdf4j.rio.RDFFormat;
+import org.eclipse.rdf4j.rio.Rio;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 20)
+@BenchmarkMode({ Mode.AverageTime })
+@Fork(value = 1, jvmArgs = { "-Xms1G", "-Xmx1G", "-XX:+UseSerialGC" })
+//@Fork(value = 1, jvmArgs = {"-Xms8G", "-Xmx8G", "-XX:+UseSerialGC", "-XX:+UnlockCommercialFeatures", "-XX:StartFlightRecording=delay=5s,duration=120s,filename=recording.jfr,settings=profile", "-XX:FlightRecorderOptions=samplethreads=true,stackdepth=1024", "-XX:+UnlockDiagnosticVMOptions", "-XX:+DebugNonSafepoints"})
+@Measurement(iterations = 10)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+public class IsomorphicBenchmark {
+
+	@Benchmark
+	public void empty() throws IOException {
+
+		isomorphic("empty.ttl");
+
+	}
+
+	@Benchmark
+	public void blankNodes() throws IOException {
+
+		isomorphic("blankNodes.ttl");
+
+	}
+
+	@Benchmark
+	public void shacl() throws IOException {
+
+		isomorphic("shacl.ttl");
+
+	}
+
+	@Benchmark
+	public void longChain() throws IOException {
+
+		isomorphic("longChain.ttl");
+
+	}
+
+	@Benchmark
+	public void sparqlTestCase() throws IOException {
+
+		isomorphic("sparqlTestCase.ttl");
+
+	}
+
+	@Benchmark
+	public void bsbm() throws IOException {
+
+		isomorphic("bsbm-100.ttl");
+
+	}
+
+	@Benchmark
+	public void spinFullForwardchained() throws IOException {
+
+		isomorphic("spin-full-forwardchained.ttl");
+
+	}
+
+	@Benchmark
+	public void list() throws IOException {
+
+		isomorphic("list.ttl");
+
+	}
+
+	@Benchmark
+	public void internallyIsomorphic() throws IOException {
+
+		isomorphic("internallyIsomorphic.ttl");
+
+	}
+
+	@Benchmark
+	public void manyProperties() throws IOException {
+
+		isomorphic("manyProperties.ttl");
+
+	}
+
+	@Benchmark
+	public void manyProperties2() throws IOException {
+
+		isomorphic("manyProperties2.ttl");
+
+	}
+
+	private Model getModel(String name) throws IOException {
+		try (InputStream resourceAsStream = IsomorphicBenchmark.class.getClassLoader()
+				.getResourceAsStream("benchmark/" + name)) {
+			return Rio.parse(resourceAsStream, "http://example.com/", RDFFormat.TURTLE);
+		}
+	}
+
+	private void isomorphic(String name) throws IOException {
+		Model m1 = getModel(name);
+		Model m2 = getModel(name);
+
+		boolean isomorphic = Models.isomorphic(m1, m2);
+		if (!isomorphic) {
+			throw new IllegalStateException("Not isomorphic");
```

Comment:
Would it be helpful to also have a few test cases for two models that are _not_ isomorphic? 

We can add those later and then adapt this I guess.

Reply:
Would be helpful.

Reply:
Would also help to benchmark early termination. Eg. when two models are obviously different. I wonder if we can do something simple like m1.size() != m2.size()?

Reply:
this is done